### ISSUE TITLE
Is String not implemented for FromSql?

### ISSUE THREAD

**Issue by dylan-bowker** [prisma/tiberius#114]
```
    |
367 |             my_string_field: row.get(12),
    |                           ^^^ the trait `FromSql<'_>` is not implemented for `std::string::String`
```

`my_string_field` is a struct field of `String` type. `Row::get()` only seems to fail on `String` types. Other fields are OK. I'm pretty newb to Rust so any input would be appreciated. As far as I know, I wouldn't be able to implement this myself because both are foreign types?

**Comment by dylan-bowker** [comment]
I'm a little bit lost here. Maybe I am doing something wrong? Maybe I can help you update your examples if you help me out here.

```
    let a: Decimal = dec!(1.23);
    let b: u32 = 123;
    let c: String = "Test 1 2 3".to_string();

    let results = client.execute(
        "INSERT INTO ##Test (id) VALUES (@P1), (@P2), (@P3)",
        &[&a, &2i32, &b, &c],
    ).await?;
```

```
error[E0277]: the trait bound `u32: ToSql` is not satisfied
  --> src\main.rs:38:22
   |
38 |         &[&a, &2i32, &b, &c],
   |                      ^^ the trait `ToSql` is not implemented for `u32`
   |
   = note: required for the cast to the object type `dyn ToSql`
```

**Comment by pimeys** [comment]
Hey! There are no stupid questions and this thing is definitely tricker due to how we try to make the API to not be slow without the user noticing it, and how the methods you use try to be as generic as possible.

Now in your case you get an instance of `Row`. You can try to get column data out from the row, but the methods `get` and `try_get` take the row by reference, meaning whatever data they return they must pass a reference or if it's cheap (such as by-value types, meaning integers, booleans and other cheap-to-copy things) you get a value back.

Now let's see how `Row::get` works:

```rust
pub fn get<'a, R, I>(&'a self, idx: I) -> Option<R>
where
    R: FromSql<'a>,
    I: QueryIdx,
{
    self.try_get(idx).unwrap()
}
```

It's a quite generic method as you can see. It will return an optional R, where R is a type that implements the trait `FromSql` and it takes an index parameter that should implement the trait `QueryIdx`.

Now, the first thing is the result of `Option`. The typing cannot be sure is the data null or not, thus relying on `Option` and letting the user to take control on how to handle nulls. Let's see how we could get a `String` from a row:

```rust
let s: Option<String> = row.get::<&str, usize>(0).map(|s| s.to_string());
```

That looks quite complex, doesn't it? Well, there are reasons why it must work like this:

- The `get` takes two type parameters (and a lifetime parameter, but we don't need to write it here): `R` and `I`. We must write them both, if we write one of them. Here we say to the `get` please get the `FromSql` for `&str`, and we take the value from the `Row` with an `usize` parameter (not, for example with `&str` which would point to the name of the column).
- The `get` returns an `Option<&str>`, so we call `Option::map` to convert the `&str` to a `String`. This operation is expensive, and the API of Tiberius makes it explicit for the user how there's a price to pay.

If the code you're writing is performance-critical, you might want to try a different approach:

```rust
for column in row.into_iter() {
    match column {
        ColumnData::String(Some(cow)) => {
            let s: String = cow.into_owned();
            ...
        }
        ...
    }
}
```

For your second question: The `ToSql` implementations are only for types that have counterparts in SQL Server that can be easily converted from Rust. For numbers this means the following:

|Rust type|Server type|
|---------|-----------|
|`u8`     |`tinyint`  |
|`i16`    |`smallint` |
|`i32`    |`int`      |
|`i64`    |`bigint`   |
|`f32`    |`float(24)`|
|`f64`    |`float(53)`|

You're passing a `u32` which doesn't have a counterpart in SQL Server. You should first convert the variable to a type matching the type in the column, handling possible overflows if needed.

Hope this was helpful, I'll try to find more time to answer if something was left unclear!

**Comment by pimeys** [comment]
It would be quite cool to have Serde specialization in Tiberius. I've been thinking about building one, but also I'm kind of expecting that to be a job for higher level database crates, such as Diesel, if they'd support SQL Server and would use Tiberius.

**Comment by dylan-bowker** [comment]
Wow I was not expecting a response this thorough. 

Mapping the `&str` to a `String` is a clever. Rust is pretty tricky. When I think of `map()` I think of an iterator. Thank you! 

I did not even consider for a moment that Microsoft SQL didn't support an unsigned 32-bit integer type! Unexpected.

Your reply was very helpful and I was able to continue my work. 

I will try my hand at writing some more examples for Tiberius on the weekend, from a newbs perspective. Hopefully I can pay your help forward.
